src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets

```diff
@@ -330,7 +330,7 @@ The .NET Foundation licenses this file to you under the MIT license.
     </ItemGroup>
     <ItemGroup Condition="'$(TargetOS)' != 'windows' and '$(TargetOS)' != 'osx'">
       <CustomLinkerArg Include="-Wl,--discard-all" />
-      <CustomLinkerArg Include="-Wl,--gc-sections" />
+      <CustomLinkerArg Include="-Wl,--gc-sections" Condition="'$(UseLLVMLinker)' != 'true'" />
```

Comment:
With `--gc-sections`, it was running into this error:

```
ld.lld: error: undefined symbol: __stop___modules
>>> referenced by main.cpp:0 (/runtime/src/coreclr/nativeaot/Bootstrap/main.cpp:0)
>>>               main.cpp.o:(InitializeRuntime()) in archive /runtime/artifacts/bin/coreclr/linux.arm64.Debug/aotsdk/libbootstrapper.a
>>> referenced by main.cpp:0 (/runtime/src/coreclr/nativeaot/Bootstrap/main.cpp:0)
>>>               main.cpp.o:(InitializeRuntime()) in archive /runtime/artifacts/bin/coreclr/linux.arm64.Debug/aotsdk/libbootstrapper.a

ld.lld: error: undefined symbol: __start___modules
>>> referenced by main.cpp:0 (/runtime/src/coreclr/nativeaot/Bootstrap/main.cpp:0)
>>>               main.cpp.o:(InitializeRuntime()) in archive /runtime/artifacts/bin/coreclr/linux.arm64.Debug/aotsdk/libbootstrapper.a
>>> referenced by main.cpp:0 (/runtime/src/coreclr/nativeaot/Bootstrap/main.cpp:0)
>>>               main.cpp.o:(InitializeRuntime()) in archive /runtime/artifacts/bin/coreclr/linux.arm64.Debug/aotsdk/libbootstrapper.a
>>> referenced by main.cpp:0 (/runtime/src/coreclr/nativeaot/Bootstrap/main.cpp:0)
>>>               main.cpp.o:(InitializeRuntime()) in archive /runtime/artifacts/bin/coreclr/linux.arm64.Debug/aotsdk/libbootstrapper.a
>>> referenced 1 more times
>>> the encapsulation symbol needs to be retained under --gc-sections properly; consider -z nostart-stop-gc (see https://lld.llvm.org/ELF/start-stop-gc)
```

The differences between GNU ld and LLVM lld are described in this article: https://maskray.me/blog/2020-12-19-lld-and-gnu-linker-incompatibilities. The weak sections are discarded by `--gc-sections` which are needed by live-sections and we don't get the `DT_NEEDED`.

Reply:
I am a bit worried that not performing `--gc-sections` might result in a size regression. I wonder if we could do something along the lines the error message above suggests:
`the encapsulation symbol needs to be retained under --gc-sections properly; consider -z nostart-stop-gc (see https://lld.llvm.org/ELF/start-stop-gc)`.

Reply:
Here are some size reduction comments from the PR in which the `--gc-sections` was added:
https://github.com/dotnet/corert/pull/7929#issuecomment-569738968

Reply:
Actually, the size reduction just from the `--gc-sections` seems to be relatively small. Another comment in that PR has more detailed overview:
https://github.com/dotnet/corert/pull/7929#issuecomment-569999179

Reply:
Ok, thank you for the details!